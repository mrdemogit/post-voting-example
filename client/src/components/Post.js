import React, { memo } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { equals } from 'ramda'
import Comment from './Comment'
import VoteSection from './VoteSection'
import { PostType } from '../types'

const StyledPost = styled.div`
  padding: 0.8em;
  margin: 0.3em;
  background: #0075a1d4;
  border-radius: 1em;
  color: #fff;
`

const StyledName = styled.div`
  color: #000;
  font-style: italic;
`

export const Post = ({
  uuid,
  author,
  content,
  votes,
  comments,
  upvote,
  downvote
}) => (
  <StyledPost>
    <StyledName>{author}</StyledName>
    <div>{content}</div>
    <VoteSection
      uuid={uuid}
      upvote={upvote}
      downvote={downvote}
      votes={votes}
    />
    {comments.map((comment) => (
      <Comment
        key={comment.uuid}
        {...comment}
        upvote={upvote}
        downvote={downvote}
      />
    ))}
  </StyledPost>
)

Post.propTypes = {
  ...PostType,
  upvote: PropTypes.func.isRequired,
  downvote: PropTypes.func.isRequired
}

const shouldUpdate = (prev, next) => {
  return equals(prev, next)
}

export default memo(Post, shouldUpdate)
