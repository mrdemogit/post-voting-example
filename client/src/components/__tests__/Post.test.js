import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { Post } from '../Post'

describe('Post', () => {
  it('should render correctly', () => {
    const renderer = new ShallowRenderer()
    renderer.render(
      <Post
        uuid='uuid1'
        author='authorTest'
        content='contentTest'
        votes={123}
        upvote={() => {}}
        downvote={() => {}}
        comments={[
          {
            uuid: 'uuid2',
            author: 'author2',
            content: 'contentTest2',
            votes: -331
          },
          {
            uuid: 'uuid3',
            author: 'author3',
            content: 'contentTest3',
            votes: 0
          }
        ]}
      />
    )

    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
})
