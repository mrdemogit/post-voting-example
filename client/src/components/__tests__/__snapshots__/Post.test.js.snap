// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post should render correctly 1`] = `
<ForwardRef>
  <ForwardRef>
    authorTest
  </ForwardRef>
  <div>
    contentTest
  </div>
  <VoteSection
    downvote={[Function]}
    upvote={[Function]}
    uuid="uuid1"
    votes={123}
  />
  <UNDEFINED
    author="author2"
    content="contentTest2"
    downvote={[Function]}
    upvote={[Function]}
    uuid="uuid2"
    votes={-331}
  />
  <UNDEFINED
    author="author3"
    content="contentTest3"
    downvote={[Function]}
    upvote={[Function]}
    uuid="uuid3"
    votes={0}
  />
</ForwardRef>
`;
